File: leetcode/recursion-backtracking/0051_n_queens.cpp
// LC 51. N-Queens | Hard
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
    unordered_set<int> cols, diag, anti;
    void bt(int row, int n, vector<string>& board, vector<vector<string>>& res) {
        if (row == n) { res.push_back(board); return; }
        for (int col = 0; col < n; col++) {
            if (cols.count(col) || diag.count(row-col) || anti.count(row+col)) continue;
            cols.insert(col); diag.insert(row-col); anti.insert(row+col);
            board[row][col] = 'Q'; bt(row+1, n, board, res); board[row][col] = '.';
            cols.erase(col); diag.erase(row-col); anti.erase(row+col);
        }
    }
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res; vector<string> board(n, string(n, '.'));
        bt(0, n, board, res); return res;
    }
};

File: leetcode/recursion-backtracking/0051_n_queens.py
# LC 51. N-Queens | Hard
class Solution:
    def solveNQueens(self, n):
        res, cols, diag, anti = [], set(), set(), set()
        board = [['.']*n for _ in range(n)]
        def bt(row):
            if row == n: res.append([''.join(r) for r in board]); return
            for col in range(n):
                if col in cols or (row-col) in diag or (row+col) in anti: continue
                cols.add(col); diag.add(row-col); anti.add(row+col)
                board[row][col] = 'Q'; bt(row+1); board[row][col] = '.'
                cols.remove(col); diag.remove(row-col); anti.remove(row+col)
        bt(0); return res
